Add information about built-in settings system (#145)
dry-rb/dry-system: docsite/source/settings.html.md

---
title: Settings
layout: gem-single
name: dry-system
---

## Basic usage

> Dry-system has already built-in settings system which you can use for your application. For this you need to use system component:

Reviewer: ```suggestion
`dry-system` provides a built-in `:settings` component which you can use in your application. In order to set it up, simply define a bootable `:settings` component and import it from the `:system` provider:
```

Reviewer 2: @solnic do you mind if we don’t use `code-style` for our gem names in docs? They’re not technically code, they’re (unconventionally lower-cased) proper nouns.

Reviewer: @timriley sure, but I cringe when I see `Dry-foo` and OTOH starting a sentence with "dry-system" looks wrong, so I dunno... 😅

Reviewer 2: I think it’s ok to start a sentence with the lower-case, normally styled text :)

Reviewer: @timriley ok great then, I'm totally ok with dropping "`" then :)

> Dry-system provides a built-in `:settings` component which you can use in your application. In order to set it up, simply define a bootable `:settings` component and import it from the `:system` provider:

Reviewer: ```suggestion
dry-system provides a built-in `:settings` component which you can use in your application. In order to set it up, simply define a bootable `:settings` component and import it from the `:system` provider:
```

Tim said it's OK to use the actual, lower-cased, gem name even in the beginning of a sentence 😄

Author: 😄

```ruby
# in system/boot/settings.rb
require "dry/system/components"
require "path/to/dry/types/file"

Application.boot(:settings, from: :system) do
  settings do
    key :database_url, Types::String.constrained(filled: true)
    key :logger_level, Types::Symbol.constructor(proc { |value| value.to_s.downcase.to_sym })
                                    .default(:info)
                                    .enum(:trace, :unknown, :error, :fatal, :warn, :info, :debug)
  end
end
```

Now, dry-system will map values from `ENV` variable to `settings` struct and allows you to use it in the application:

```ruby
Application[:settings] # => dry-struct object with settings
```

You can also use settings object in other bootable dependencies:

```ruby
Application.boot(:redis) do |container|
  init do
    use :settings

    uri = URI.parse(container[:settings].redis_url)
    redis = Redis.new(host: uri.host, port: uri.port, password: uri.password)

    container.register('persistance.redis', redis)
  end
end
```

> Or use settings in your logic:

Reviewer: ```suggestion
Or use `:settings` as an injectible dependency in your classes:
```

```ruby
  module Operations
    class CreateUser
      include Import[:settings, :repository]

      def call(id:)
        settings # => dry-struct object with settings

        # ...
      end
    end
  end
end
```

## Default values

You can [use dry-type](https://dry-rb.org/gems/dry-types/master/default-values/) for provide default value for specific setting:

```ruby
settings do
  key :redis_url, Types::Coercible::String.default('')
end
```

> In this case, if you don't have `ENV['REDIS_URL']` value you get `''` as a default value for `settings.redis_url` calls

Reviewer: ```suggestion
In this case, if you don't have `ENV['REDIS_URL']` value, you get `''` as the default value for `settings.redis_url` calls.
```
